Sources/NIOCore/ByteBuffer-aux.swift

```diff
@@ -125,6 +137,18 @@ extension ByteBuffer {
             return self._setStringSlowpath(string, at: index)
         }
     }
+    
+    /// Write `string` null terminated into this `ByteBuffer` at `index` using UTF-8 encoding. Does not move the writer index.
+    ///
+    /// - parameters:
+    ///     - string: The string to write.
+    ///     - index: The index for the first serialized byte.
+    /// - returns: The number of bytes written.
+    public mutating func setNullTerminatedString(_ string: String, at index: Int) -> Int {
+        return string.utf8CString.withUnsafeBytes {
```

Comment:
@Lukasa did you check that this allocates? Because the idea with Swift 5(?)'s (then) new String was that we always store it 0 terminated so this can be free. So I'd expect that the `utf8CString` view is the same as the `utf8` view but with the extra 0 byte visible.

Happy to be wrong but if I'm wrong we should file this, no need for Swift to alloc here.

Reply:
`utf8CString` returns `ContiguousArray<CInt>`. This does not match any of the internal views of the String, and so it allocates.

Reply:
OMG, thanks, TIL this returns a `ContiguousArray`!?. I think this is still worthy a Swift bug (CC @milseman). IMHO, properties should be _O(1)_ and I don't see why we can't make a view here much like `UTF8View` that we can get a contiguous pointer on.

Reply:
That's not from Swift, it's from Foundation.

Reply:
Sure seems like it's from Swift: 

https://github.com/apple/swift/blob/17f1cf92973d2c1b3d5fe6b70f4a33771a303e3a/stdlib/public/core/StringUTF8View.swift#L253-L264

Reply:
We copy the contents anyway so why not just write the string and append `null` at the end? We would just need to reserve `string.utf8.count + 1` bytes upfront to not trigger a realloc.

Reply:
@dnadoba yes, we can do that but the stdlib stores 0 terminated anyway (except smol ones) so you'd expect that it's able to vend us such a pointer (without allocing) :)

Reply:
Yeah, my advice to @fabianfett was to simply write the string and append a null. The performance delta won't be worth it.

Reply:
@Lukasa ah yes, that's the one. I got it confused with `utf8String`, which likely has a signature you'd prefer. `utf8CString` is an old API and I'm ok with deprecating it in favor of something else. What would that type be, given you want separate ownership?

Reply:
I think @weissi proposed having a view type (like `UTF8View` that was null-terminated). To allow separate ownership we probably need the view to be wide, so that it can store a smol string with the null-terminator. That should allow zero allocations to produce.

Reply:
We model shared ownership via closures today, ala `withCString` which will only allocate for non-native strings. Native strings are tail-allocated. When we get move only types, we can have move-only shared views and partially-escape closure hell.

Reply:
How does `UTF8View` fit into that model?

Reply:
UTFView doesn't unconditionally provide access to contiguous UTF-8 encoded bytes in memory. You can conditionally get those if available via `wCSIA`